Firestore/Source/API/FIRDocumentSnapshot.mm

```diff
@@ -88,79 +90,55 @@ + (instancetype)snapshotWithFirestore:(FIRFirestore *)firestore
 @end
 
 @implementation FIRDocumentSnapshot {
-  FIRSnapshotMetadata *_cachedMetadata;
-  DocumentKey _internalKey;
+  DocumentSnapshot _shim;
 }
 
-@dynamic metadata;
-
 - (instancetype)initWithFirestore:(FIRFirestore *)firestore
```

Comment:
Since the guts of this type are now contained in the `DocumentSnapshot`, could we make this initializer be `initWithSnapshot:(DocumentSnapshot &&)snapshot` instead?

That would make the factory method `snapshotWithSnapshot:(DocumentSnapshot)snapshot` too and drastically cut down on the verbosity of all this parameter passing.

Note:
  * We use this factory method technique here to hide the initializer from the public interface. You can add arbitrary methods but you can't add designated initializers.
  * The initializer can take a move-only parameter because the factory shouldn't ever do anything but move.

Reply:
Done. Note that I also implemented `QueryDocumentSnapshot` now, so that it now takes a (unique) pointer to `DocumentSnapshot`.